specs/core/0_beacon-chain.md

```diff
@@ -998,28 +1032,42 @@ def get_block_root(state: BeaconState,
 
 ```python
 def get_randao_mix(state: BeaconState,
-                   slot: SlotNumber) -> Bytes32:
+                   epoch: EpochNumber) -> Bytes32:
     """
     Returns the randao mix at a recent ``slot``.
     """
-    assert state.slot < slot + LATEST_RANDAO_MIXES_LENGTH
-    assert slot <= state.slot
-    return state.latest_randao_mixes[slot % LATEST_RANDAO_MIXES_LENGTH]
+    assert get_current_epoch(state) < epoch + LATEST_RANDAO_MIXES_LENGTH
+    assert epoch <= get_current_epoch(state)
+    return state.latest_randao_mixes[epoch % LATEST_RANDAO_MIXES_LENGTH]
 ```
 
 #### `get_active_index_root`
 
 ```python
 def get_active_index_root(state: BeaconState,
-                          slot: SlotNumber) -> Bytes32:
+                          epoch: EpochNumber) -> Bytes32:
     """
-    Returns the index root at a recent ``slot``.
+    Returns the index root at a recent ``epoch``.
     """
-    state_epoch = state.slot // EPOCH_LENGTH
-    given_epoch = slot // EPOCH_LENGTH 
-    assert state_epoch < given_epoch + LATEST_INDEX_ROOTS_LENGTH
-    assert given_epoch <= state_epoch
-    return state.latest_index_roots[given_epoch % LATEST_INDEX_ROOTS_LENGTH]
+    assert get_current_epoch(state) < epoch + LATEST_INDEX_ROOTS_LENGTH
+    assert epoch <= get_current_epoch(state)
+    return state.latest_index_roots[epoch % LATEST_INDEX_ROOTS_LENGTH]
+```
+
+#### `generate_seed`
+
+```python
+def generate_seed(state: BeaconState,
+                  epoch: EpochNumber) -> Bytes32:
+    """
+    Generate a seed for the given ``epoch``.
+    """
+    randao_mix_epoch = epoch + LATEST_RANDAO_MIXES_LENGTH - SEED_LOOKAHEAD
```

Comment:
Huh? Why `+ LATEST_RANDAO_MIXES_LENGTH`?

If it's to avoid underflow IMO this is going too far and sacrifices clarity; we should have the anti-underflow logic in the `get_randao_mix` function.

Reply:
Ah, the epoch starting at current unix time will fix this. Making an issue for that now

Reply:
do we want to change this in this PR?